libpdb/topic.go

```diff
@@ -91,105 +104,74 @@ func (t *Topic) GeneratePublish(commonConfig *common.CommonConfig, seqNo uint64,
-		t.drbg.FillBytes(args[1].ForTd[j].RequestVector)
-		args[1].ForTd[j].PadSeed = make([]byte, drbg.SeedLength)
-		t.drbg.FillBytes(args[1].ForTd[j].PadSeed)
-	}
-	// @todo - XOR this into the last request vector
-	//k0, k1 = t.Seed2.KeyUint128()
-	//bucket2 := siphash.Hash(k0, k1, seqNoBytes) % globalConfig.NumBuckets
-
-	return args[0], args[1], nil
+// @TODO: signing. likely via https://github.com/agl/ed25519
+// @TODO: long-term, keys should ratchet, so that messages outside of the
+// active range become refutable. perhaps this could alternatively be done with
+// a server managed primative, with releases of a rachet update as each DB epoch
+// advances.
+func (t *Topic) encrypt(plaintext []byte, nonce *[24]byte) ([]byte, error) {
+  buf := make([]byte, 0, len(plaintext) + box.Overhead)
+	_ = box.SealAfterPrecomputation(buf, plaintext, nonce, t.sharedSecret)
+	buf = buf[0:cap(buf)]
+	digest := ed25519.Sign(t.signingPrivateKey, buf)
+	return append(buf, digest[:]...), nil
 }
 
-func (t *Topic) retrieveResponse(args *common.ReadArgs, reply *common.ReadReply) []byte {
-	data := reply.Data
-
-	for i := 0; i < len(args.ForTd); i++ {
-		pad := make([]byte, len(data))
-		seed, _ := drbg.ImportSeed(args.ForTd[i].PadSeed)
-		hashDrbg, _ := drbg.NewHashDrbg(seed)
-		hashDrbg.FillBytes(pad)
-		for j := 0; j < len(data); j++ {
-			data[j] ^= pad[j]
-		}
-	}
-	return data
-}
+type binaryTopic struct {
+	Id    uint64
+	Seed1 []byte
+	Seed2 []byte
 
-//@todo - can we use seqNo as the nonce?
-func (t *Topic) Encrypt(plaintext []byte, nonce []byte) ([]byte, error) {
-	// The key argument should be the AES key, either 16 or 32 bytes
-	// to select AES-128 or AES-256.
-	block, err := aes.NewCipher(t.EncrKey)
-	if err != nil {
-		return nil, err
-	}
+	// Keys
+	SharedSecret  [32]byte
+	SigningPrivateKey [64]byte
+	SigningPublicKey [32]byte
 
-	// Never use more than 2^32 random nonces with a given key because of the risk of a repeat.
-	/**
-	nonce := make([]byte, 12)
-	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
-		return nil, nil, err
-	}
-	**/
+	// Last seen sequence number
+	Seqno uint64
+}
 
-	aesgcm, err := cipher.NewGCM(block)
+/** Implement BinaryMarshaler / BinaryUnmarshaler for serialization **/
+func (t *Topic) MarshalBinary() (data []byte, err error) {
+	forExport := binaryTopic{t.Id, t.Seed1.Export(), t.Seed2.Export(), *t.sharedSecret, *t.signingPrivateKey, *t.signingPublicKey, t.Seqno}
+	var output bytes.Buffer
+	enc := gob.NewEncoder(&output)
+	err = enc.Encode(forExport)
 	if err != nil {
 		return nil, err
 	}
-
-	ciphertext := aesgcm.Seal(nil, nonce, plaintext, nil)
-	//fmt.Printf("%x\n", ciphertext)
-	return ciphertext, nil
+	return output.Bytes(), nil
 }
 
-func (t *Topic) Decrypt(ciphertext []byte, nonce []byte) ([]byte, error) {
-	// The key argument should be the AES key, either 16 or 32 bytes
-	// to select AES-128 or AES-256.
-	block, err := aes.NewCipher(t.EncrKey)
+func (t *Topic) UnmarshalBinary(data []byte) error {
```

Comment:
oh I see you have a binaryTopic data structure. Seems like we have a lot of redundancy in data structures...